subsystem/apiclient/kubeapi.go

```diff
@@ -0,0 +1,288 @@
+		key.Name, key.Namespace); res.Error != nil {
+		return nil, res.Error
+	}
+	return &c.Cluster, nil
+}
+
+func (kc *KubeAPIClient) DeregisterCluster(
+	ctx context.Context,
+	params *installer.DeregisterClusterParams,
+) error {
+
+	cluster, getErr := kc.getClusterByID(params.ClusterID)
+	if getErr != nil {
+		return errors.Wrapf(getErr, "failed getting cluster from db")
+	}
+	ref := &corev1.ObjectReference{
+		Name:      cluster.Name,
+		Namespace: kc.namespace,
+	}
+	if delErr := kc.client.Delete(ctx, ref); delErr != nil {
+		return errors.Wrapf(delErr, "failed deleting cluster")
+	}
+	return nil
+}
+
+func (kc *KubeAPIClient) getClusterByID(id strfmt.UUID) (*models.Cluster, error) {
+	c := &common.Cluster{}
+	if res := kc.db.Take(c, "id = ?", id.String()); res.Error != nil {
+		return nil, res.Error
+	} else if res.RowsAffected == 0 {
+		return nil, errors.New("cluster was not found in db")
+	}
+	return &c.Cluster, nil
+}
+
+func (kc *KubeAPIClient) UpdateCluster(
+	ctx context.Context,
+	params *installer.UpdateClusterParams,
+) (*models.Cluster, error) {
+
+	cParams := params.ClusterUpdateParams
+
+	clusterName := swag.StringValue(cParams.Name)
+
+	pullSecretRef, pullSecretErr := kc.deployPullSecret(
+		ctx,
+		clusterName,
+		swag.StringValue(cParams.PullSecret))
+	if pullSecretErr != nil {
+		return nil, errors.Wrapf(pullSecretErr, "failed to deploy pull secret")
+	}
+
+	metaName := uuid.New().String()
+	c := adiiov1alpha1.Cluster{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Cluster",
+			APIVersion: APIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      metaName,
+			Namespace: kc.namespace,
+		},
+		Spec: adiiov1alpha1.ClusterSpec{
+			Name:                     clusterName,
+			BaseDNSDomain:            swag.StringValue(cParams.BaseDNSDomain),
+			ClusterNetworkCidr:       swag.StringValue(cParams.ClusterNetworkCidr),
+			ClusterNetworkHostPrefix: swag.Int64Value(cParams.ClusterNetworkHostPrefix),
+			ServiceNetworkCidr:       swag.StringValue(cParams.ServiceNetworkCidr),
+			IngressVip:               swag.StringValue(cParams.IngressVip),
+			SSHPublicKey:             swag.StringValue(cParams.SSHPublicKey),
+			VIPDhcpAllocation:        swag.BoolValue(cParams.VipDhcpAllocation),
+			HTTPProxy:                swag.StringValue(cParams.HTTPProxy),
+			HTTPSProxy:               swag.StringValue(cParams.HTTPSProxy),
+			NoProxy:                  swag.StringValue(cParams.NoProxy),
+			UserManagedNetworking:    swag.BoolValue(cParams.UserManagedNetworking),
+			AdditionalNtpSource:      swag.StringValue(cParams.AdditionalNtpSource),
+			PullSecretRef:            pullSecretRef,
+		},
+	}
+	if deployErr := kc.client.Update(ctx, &c); deployErr != nil {
+		return nil, errors.Wrapf(deployErr, "failed to deploy cluster")
+	}
+
+	key := types.NamespacedName{
+		Name:      metaName,
+		Namespace: kc.namespace,
+	}
+	cluster, getClusterErr := kc.getClusterByKey(key)
```

Comment:
I think that there maybe a race here, we don't really know when the resource was updated because the API is async. 
We can take a look at `updated_at` but it's not perfect because cluster can be updated by the monitors as well.
@mhrivnak how can we validate that reconcile was called for specific resource?

Reply:
I made a general comment about this: https://github.com/openshift/assisted-service/pull/923#issuecomment-761181513